plugins/push/frontend/public/javascripts/countly.component.message-editor-with-emoji-picker.js

```diff
@@ -0,0 +1,172 @@
+(function(countlyPushNotificationComponent) {
+
+    countlyPushNotificationComponent.MessageEditorWithEmojiPicker = countlyVue.views.create({
+        template: CV.T("/push/templates/message-editor-with-emoji-picker.html"),
+        props: {
+            id: {
+                type: String,
+                required: true
+            },
+            type: {
+                type: String,
+                required: false,
+                default: "text"
+            },
+            isDefaultLocalizationActive: {
+                type: Boolean,
+                required: false,
+                default: false
+            },
+            placeholder: {
+                type: String,
+                required: false,
+                default: ""
+            }
+        },
+        data: function() {
+            return {
+                search: "",
+                defaultLabelValue: "Select property",
+                defaultLabelPreview: "Select property|",
+                defaultLocalizationValidationErrors: [],
+            };
+        },
+        computed: {
+            hasDefaultSlot: function() {
+                return Boolean(this.$slots.default);
+            },
+            hasValidationErrors: function() {
+                return this.isDefaultLocalizationActive && this.hasDefaultLocalizationValidationErrors;
+            },
+            hasDefaultLocalizationValidationErrors: function() {
+                return Boolean(this.defaultLocalizationValidationErrors.length);
+            },
+        },
+        watch: {
+            isDefaultLocalizationActive: function() {
+                this.validate();
+            }
+        },
+        methods: {
+            appendZeroWidthSpace: function() {
+                this.$refs.element.appendChild(document.createTextNode('\u200B'));
+            },
+            setCursorAtEnd: function() {
+                var selection = window.getSelection();
+                var textRange = document.createRange();
+                selection.removeAllRanges();
+                textRange.selectNodeContents(this.$refs.element);
+                textRange.collapse(false);
+                selection.addRange(textRange);
+                this.$refs.element.focus();
+            },
+            onUserPropertyClick: function(id, container, element) {
+                var elementBound = element.getBoundingClientRect();
+                var leftCoordinate = elementBound.left + (elementBound.width / 2);
+                this.$emit("click", id, container, {left: leftCoordinate, top: elementBound.top });
+            },
+            getOnUserPropertyClickEventListener: function(id, name) {
+                var self = this;
+                return function(event) {
+                    self.onUserPropertyClick(id, name, event.target);
+                };
+            },
+            addEmptyUserProperty: function(id, container) {
+                var newElement = document.createElement("span");
```

Comment:
Vue does it since it's still a component but only its value is not controlled by it. Nevertheless, I will have to double check in case I might have missed something.